### DISCUSSION TITLE
Question Regarding Subscription class and how datachange_notification works

### DISCUSSION THREAD

**Discussion by jordanjj8** [FreeOpcUa/python-opcua#1413] [Q&A] [answered]
I had a question regarding how datachange_notification works. If I have created an instance of a subscription handler that updates opcua node values, how does this work in relation to datachange_notification? The subscription takes in a period for how often it receives from the publisher, but my understanding is that datachange_notification is only triggered when a change in opcua value occurs? More clarification on this would be helpful.

Also, how does one determine if the client is getting stale opcua values in a sub-pub situation?

**Comment by AndreasHeine** [comment, marked as answer]
a opc ua subscription and pubsub are two different things!

https://reference.opcfoundation.org/v104/Core/docs/Part4/5.13.1/

 
 ​In simple words: 
 ​A OPC UA Subscription works like a "Mailbox" which gets emptied in a defined interval by the Client.  ​The OPC UA Client sends a publishrequest, the server takes all notifications since the last publishrequest and send them in the publishresponse to the client.  ​If there is no Notification in the "Mailbox" the Client will get a "keep-alive" back. ​The OPC UA Client can add "MonitoredItems" to the Subscription which will generate Notifications.

**Reply by jordanjj8** [reply]
Very much appreciate the response and link. And yes, I mean using OPC UA subscription in context of client to server, not pubsub. 


I am currently looking at 
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.12.1/

In figure 15, it says "the sampling interval defines the cyclic rate used by the server to sample the real item". And in 5.12.1.2 it says "Each MonitoredItem created by the Client is assigned a sampling interval that is either inherited from the publishing interval of the Subscription or that is defined specifically to override that rate".

Q: How does one change the sampling interval? Is this the sub period?

Also regarding the queue size:
"The size of the queue is defined when the MonitoredItem is created. When the queue is full and a new Notification is received, the Server either discards the oldest Notification and queues the new one, or it replaces the last value added to the queue with the new one."

Q: Where can the queue size be set? On the client or server side? If on the client, where?

**Reply by jordanjj8** [reply]
Also in your mailbox analogy, is the mail the publishrepsonse?

**Reply by AndreasHeine** [reply]
the publishresponse is a package of all letters in the mailbox!

**Reply by AndreasHeine** [reply]
```python
    node_handles = subscription.subscribe_data_change(
        nodes=nodes, # a list of nodes i want to subscribe to
        attr=ua.AttributeIds.Value, # the attribute i am interested in
        queuesize=50, # the queuesize should be bigger then the number of changes within a publishinterval, in this case 50 valuechanges per 1000 ms
        monitoring=ua.MonitoringMode.Reporting
    )
```

**Reply by jordanjj8** [reply]
Thank you @AndreasHeine , how about the sampling interval?

**Reply by jordanjj8** [reply]
also "in this case 50 valuechanges per 1000 ms", where is the 1000 ms set?

**Reply by AndreasHeine** [reply]
> also "in this case 50 valuechanges per 1000 ms", where is the 1000 ms set?

inherit from the publishinterval!

if you want full control you will need to make a lowlevel call to create the monitored item!

https://github.com/FreeOpcUa/python-opcua/blob/9c1268574402f69d00c835d0f9adac3bd951509a/opcua/ua/uaprotocol_auto.py#L8636

```python
        self.ClientHandle = 0
        self.SamplingInterval = 0 # <---------- !!!
        self.Filter = ExtensionObject()
        self.QueueSize = 0
        self.DiscardOldest = True
```

`MonitoredItemCreateRequest.RequestedParameters <- MonitoringParameters`

https://github.com/FreeOpcUa/python-opcua/blob/9c1268574402f69d00c835d0f9adac3bd951509a/opcua/ua/uaprotocol_auto.py#L8672

`self.ItemsToCreate = [] <- [MonitoredItemCreateRequest]`

https://github.com/FreeOpcUa/python-opcua/blob/9c1268574402f69d00c835d0f9adac3bd951509a/opcua/common/subscription.py#L257

**Comment by jordanjj8** [comment]
@AndreasHeine To ensure that the queue actually gets emptied, and we get the most up to date, should we have the subscribing rate be higher than the publishing rate from the server? 

Also, what is the perceived benefit of having a queue size greater than 0? Would this even be a benefit if we can ensure that the subscribing rate is higher than the publishing rate?

**Reply by AndreasHeine** [reply]
the queue just have to be bigger then the number of valuechanges you got. -> then you will not miss any valuechange! if you have limited cpu/ram (small devices) you cant use a big queue and long publishintervals. you need to decide... 

the queue is empty after each publishresponse (if it fits in one msg)!

the publishinterval is just the delay in which you get data but it does not mean you miss any changes! having a proper queue makes the length of the publishresponse predictable!

**Reply by AndreasHeine** [reply]
Sampling -> https://reference.opcfoundation.org/v104/Core/docs/Part4/5.12.1/

**Comment by jordanjj8** [comment]
@AndreasHeine 
> the queue is empty after each publishresponse (if it fits in one msg)!

This doesn't make sense to me. If lets say the queue of a node value, that is a bool, is 10 messages, how can a value be a list of 10 messages if its supposed to be either true or false? 

Another way to put my question: If after each publishresponse the client receives all notifications from the queue for the MonitoredItem (in this case a bool) to the Client, how does the client process these values?

**Reply by AndreasHeine** [reply]
> This doesn't make sense to me. If lets say the queue of a node value, that is a bool, is 10 messages, how can a value be a list of 10 messages if its supposed to be either true or false?

remember a DataChangeNotification is always a Value or StatusCode change! And the Queue just has to buffer the messages till next publishrequest or while publishing is disabled
https://reference.opcfoundation.org/src/v104/Core/docs/Part4/readme_files/image017.png
so if a boolean changes just 5 times in an interval (e.g. queuesize 10) you will recv 5 DataChangeNotifications -> Queue empty after publishresponse!

> Another way to put my question: If after each publishresponse the client receives all notifications from the queue for the MonitoredItem (in this case a bool) to the Client, how does the client process these values?

in order via callback!
https://github.com/FreeOpcUa/python-opcua/blob/67f15551884d7f11659d52483e7b932999ca3a75/examples/client-example.py#L22

**Reply by jordanjj8** [reply]
ah okay, so is it correct to say that the rate at which the callback is made will be the rate at which the client goes through all the DataChangeNotifications that was emptied from the queue?

Now is there a limit as to how many DataChangeNotifications can the callback handle?

**Comment by jordanjj8** [comment]
@AndreasHeine What is the difference (conceptually) between the **publish interval** and the **sampling interval**? It seems like one is specific for each MonitoredItem and the other applies generally for the subscription.

Thank you so much for your detailed answers!

**Reply by AndreasHeine** [reply]
https://documentation.unified-automation.com/uasdkhp/1.1.1/html/subscription_model.png
